agent/src/agentclient.rs

```diff
@@ -0,0 +1,173 @@
+        let associated_node_name = env::var("MY_NODE_NAME").context(error::GetNodeName)?;
+
+        let nodes: Api<Node> = Api::all(self.k8s_client.clone());
+        let associated_node: Node = nodes
+            .get(&associated_node_name)
+            .context(error::FetchNode)
+            .await?;
+
+        let associated_node_uid = match associated_node.metadata.uid {
+            Some(uid) => uid,
+            None => return error::MissNodeUid {}.fail(),
+        };
+
+        Ok(BottlerocketNodeSelector {
+            node_name: associated_node_name,
+            node_uid: associated_node_uid,
+        })
+    }
+
+    async fn get_server_domain(&self) -> &str {
+        // Todo (Tianhg)
+
+        "52.24.238.246:30000"
+    }
+
+    /// Gather metadata about the system, node status provided by spec.
+    async fn gather_system_metadata(
+        &self,
+        state: BottlerocketNodeState,
+    ) -> Result<BottlerocketNodeStatus> {
+        let current_version = get_os_info().context(error::BottlerocketNodeStatusVersion)?;
+        let update_versions =
+            list_available().context(error::BottlerocketNodeStatusAvailableVersions)?;
+
+        Ok(BottlerocketNodeStatus {
+            current_version: current_version.version_id.clone(),
+            available_versions: update_versions,
+            current_state: state,
+        })
+    }
+
+    /// update the custom resource associated with this node
+    async fn update_metadata_custom_resource(
+        &self,
+        _current_metadata: BottlerocketNodeStatus,
+    ) -> Result<()> {
+        let client = reqwest::Client::new();
+        let currernt_node_selector = self.get_node_selector().await?;
+        let server_domain = self.get_server_domain().await;
+
+        let node_req = UpdateBottlerocketNodeRequest {
+            node_status: _current_metadata,
+            node_selector: currernt_node_selector,
+        };
+
+        let response = client
+            .put(format!(
+                "http://{}/bottlerocket-node-resource",
+                &server_domain
+            ))
+            .json(&node_req)
+            .send()
+            .await
+            .context(error::UpdateBottlerocketNodeResource)?;
+
+        log::info!("{}", response.text().await.unwrap());
+
+        Ok(())
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        // A running agent has two responsibilities:
+        // - Gather metadata about the system and update the custom resource associated with this node
+        // - Determine if the spec on the system's custom resource demands the node take action. If so, begin taking that action.
+
+        loop {
+            // Requests metadata 'status' and 'spec' for the current BottlerocketNode
+            let bottlerocket_node_status = match self.fetch_custom_resource().await?.status {
+                Some(status) => status,
+                None => return error::MissBottlerocketNodeStatus {}.fail(),
+            };
```

Comment:
If you include `snafu::OptionExt`, you can do:

```suggestion
            let bottlerocket_node_status = self
              .fetch_customer_resource()
              .await?
              .status
              .context(error::MissBottlerocketNodeStatus)?;
```